wgengine/wgtun/tun.go

```diff
@@ -0,0 +1,146 @@
+// Copyright (c) 2020 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package wgtun
+
+import (
+	"errors"
+	"os"
+	"sync"
+
+	"github.com/tailscale/wireguard-go/device"
+	"github.com/tailscale/wireguard-go/tun"
+	"tailscale.com/types/logger"
+	"tailscale.com/wgengine/filter"
+)
+
+const MaxPacketSize = device.MaxContentSize
+
+var (
+	ErrFiltered     = errors.New("packet dropped by filter")
+	ErrPacketTooBig = errors.New("packet too big")
+)
+
+// Device wraps a tun.Device from wireguard-go,
+// augmenting it with filtering and packet injection.
+// All the added work happens in Read and Write:
+// the other methods delegate to the underlying innerDevice.
+type Device struct {
+	innerDevice tun.Device
+
+	// Inbound packets can be written directly into the TUN device,
+	// but the other direction must wait on Wireguard to poll it.
+	// The packets are kept in the queue while waiting.
+	outbound chan []byte
+
+	// Filters are updated by ipn/local during the netmap refresh.
+	// They may be in use when this happens, hence the need for locking.
+	filterLock sync.Mutex
+	filterIn   func(b []byte) filter.Response
+	filterOut  func(b []byte) filter.Response
+}
+
+func WrapTun(logf logger.Logf, device tun.Device) *Device {
+	nofilter := func(b []byte) filter.Response {
+		// for safety, default to dropping all packets
+		logf("Warning: you forgot to use tundev.SetFilterInOut()! Packet dropped.")
+		return filter.Drop
+	}
+
+	return &Device{
+		innerDevice: device,
+		outbound:    make(chan []byte),
+		filterIn:    nofilter,
+		filterOut:   nofilter,
+	}
+}
+
+func (d *Device) Close() error {
+	return d.innerDevice.Close()
+}
+
+func (d *Device) Events() chan tun.Event {
+	return d.innerDevice.Events()
+}
+
+func (d *Device) File() *os.File {
+	return d.innerDevice.File()
+}
+
+func (d *Device) Flush() error {
+	return d.innerDevice.Flush()
+}
+
+func (d *Device) MTU() (int, error) {
+	return d.innerDevice.MTU()
+}
+
+func (d *Device) Name() (string, error) {
+	return d.innerDevice.Name()
+}
+
+func (d *Device) Read(buf []byte, offset int) (n int, err error) {
+	select {
+	case packet := <-d.outbound:
+		n = copy(buf[offset:], packet)
+	default:
+		n, err = d.innerDevice.Read(buf, offset)
```

Comment:
If this blocks, what happens if somebody sends to d.outbound? Should this Read happen in an another goroutine and send to outbound? (like magicsock does for UDP packets coming from UDP+DERP)

Reply:
You're right, unfortunately Windows does block. Working around this increases the complexity of the solution significantly, so I'll think about it some more.